### DISCUSSION TITLE
Help w. script for switch

### DISCUSSION THREAD

**Discussion by fabioeidi20** [RavenSystem/esp-homekit-devices#2032] [MEPLHAA Scripts]
Back in version 11.x I had this script:

{"c":{"l":12,"n":"HAA-CAA776","h":1,"b":[{"g":3,"t":5}]},"a":[{"t":2,"0":{"r":[{"g":14,"v":0},{"g":13,"v":1}]},"1":{"r":[{"g":14,"v":1},{"g":13,"v":0}]},"b":[{"g":3,"t":1}]}]}

In v. 12 it no longer works due to the IO stuff.

I have a new script, and managed to figure out that GPIOs 12 and 13 are LEDs so mode 2 worked well.

But I can't seem to figure out the mode for GPIO14, which controls the actual relay inside the switch. Any help? I already tried mode 6 but it aint it.

**Comment by mario-website** [comment]
Could you try to use my converter:
https://converter-to-merlin.netlify.app/
In your case should working fine and let me know if is working or not.

**Reply by pasparakis** [reply]
Very good idea the converter..but not working for me..😐

**Reply by mario-website** [reply]
> Very good idea the converter..but not working for me..😐

I learning Merlin... I believe this is related to https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4

**Reply by mario-website** [reply]
Anyway, could you show me your code where is working GPIO 12 and 13?

**Reply by xrust83** [reply]
Almost any old Sonoff device (Basic, TH, mini, switch…) Gpio13 LED, Gpio12 Relay.

Sonoff Basic
{"c":{"io":[[[12,13],2],[[0,5],6]],"l":13,"n":"Bathroom-Light","b":[{"g":0,"t":5}]},"a":[{"0":{"r":[{"g":12}]},"1":{"r":[{"g":12,"v":1}]},"b":[{"g":5},{"g":5,"t":0}]}]}

Sonoff mini - 
{"c":{"io":[[[12,13],2],[[0,4],6]],"l":13,"b":[{"g":0,"t":5}],"n": "Entrance-Light"},"a":[{"0":{"r":[{"g":12}]},"1":{"r":[{"g":12,"v":1}]},"b":[{"g":0},{"g":4},{"g":4,"t":0}]}]}

Sonoff TH -
{"c":{"io":[[[12,13],2],[[0],6]],"l":13,"ct":7,"b":[{"g":0,"t":5}],"n":"Atlantic-1500"},"a":[{"0":{"r":[{"g":12,"v":0}]},"1":{"r":[{"g":12,"v":1}]},"t":2,"b":[{"g":0}],"s":5,"es":[{"t":22,"g":14,"n":3,"b":[],"j":5,"z":15}]}]}

**Reply by mario-website** [reply]
> Almost any old Sonoff device (Basic, TH, mini, switch…) Gpio13 LED, Gpio12 Relay.
> 
> Sonoff Basic {"c":{"io":[[[12,13],2],[[0,5],6]],"l":13,"n":"Bathroom-Light","b":[{"g":0,"t":5}]},"a":[{"0":{"r":[{"g":12}]},"1":{"r":[{"g":12,"v":1}]},"b":[{"g":5},{"g":5,"t":0}]}]}
> 
> Sonoff mini - {"c":{"io":[[[12,13],2],[[0,4],6]],"l":13,"b":[{"g":0,"t":5}],"n": "Entrance-Light"},"a":[{"0":{"r":[{"g":12}]},"1":{"r":[{"g":12,"v":1}]},"b":[{"g":0},{"g":4},{"g":4,"t":0}]}]}
> 
> Sonoff TH - {"c":{"io":[[[12,13],2],[[0],6]],"l":13,"ct":7,"b":[{"g":0,"t":5}],"n":"Atlantic-1500"},"a":[{"0":{"r":[{"g":12,"v":0}]},"1":{"r":[{"g":12,"v":1}]},"t":2,"b":[{"g":0}],"s":5,"es":[{"t":22,"g":14,"n":3,"b":[],"j":5,"z":15}]}]}

So you want me to tell me that configuration file after conversion will looks like:
{"c":{"l":12,"n":"HAA-CAA776","h":1,"b":[{"g":3,"t":5}],"io":[[[3],6],[[14,13,12],2]]},"a":[{"0":{"r":[{"g":14,"v":0},{"g":13,"v":1}]},"1":{"r":[{"g":14,"v":1},{"g":13,"v":0}]},"t":2,"b":[{"g":3,"t":1}]}]}
?

**Reply by xrust83** [reply]
sonoff work with configuration "io":[[[12,13],2]] and configurations “io”:[[[12],2],[[13],3]] 
In the beta version, Jose did an example as in the last configuration, but then he redirected it as in the first one.

I think this is due to the way the LED is turned on according to the scheme (without a transistor, with a field-effect transistor or a bipolar transistor), so the configurator is not always the right solution. The same situation is with the pull-up resistors of the buttons. Their presence or absence must be determined physically, and from this a decision should be made [0],6 or [0],6,1. All Sonoff [0],6 because the resistor is provided on the Itead boards

**Reply by fabioeidi20** [reply]
> > Very good idea the converter..but not working for me..😐
> 
> I learning Merlin... I believe this is related to https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4

This converter works for me! Thanks for the help!

**Reply by xrust83** [reply]
> > > Very good idea the converter..but not working for me..😐
> > 
> > 
> > I learning Merlin... I believe this is related to https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4
> 
> This converter works for me! Thanks for the help!

{"c":{"l":12,"n":"HAA-CAA776","h":1,"b":[{"g":3,"t":5}],"io":[[[3],6],[[14,13],2]]},"a":[{"0":{"r":[{"g":14,"v":0},{"g":13,"v":1}]},"1":{"r":[{"g":14,"v":1},{"g":13,"v":0}]},"t":2,"b":[{"g":3,"t":1}]}]} 

Yes, I also tried to convert. currently your status LED is not working because it is not declared in “io”: Right?

**Reply by fabioeidi20** [reply]
> > > > Very good idea the converter..but not working for me..😐
> > > 
> > > 
> > > I learning Merlin... I believe this is related to https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4
> > 
> > 
> > This converter works for me! Thanks for the help!
> 
> {"c":{"l":12,"n":"HAA-CAA776","h":1,"b":[{"g":3,"t":5}],"io":[[[3],6],[[14,13],2]]},"a":[{"0":{"r":[{"g":14,"v":0},{"g":13,"v":1}]},"1":{"r":[{"g":14,"v":1},{"g":13,"v":0}]},"t":2,"b":[{"g":3,"t":1}]}]}
> 
> Yes, I also tried to convert. currently your status LED is not working because it is not declared in “io”: Right?

The LED works when 12 and 13 are declared in "io" as mode 2, like in your example. 

Thanks for the help!

**Reply by mario-website** [reply]
So "l": VALUE should also go to array where is [[VALUE],2], right? If yes, will update converter.

**Comment by pasparakis** [comment]
last night the configurator worked..but today it doesn't work..it's very convenient if it works..thanks..16 Μαΐ 2023, 19:18, ο χρήστης «mario-website ***@***.***>» έγραψε:﻿

Very good idea the converter..but not working for me..😐

I learning Merlin... I believe this is related to [https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4]

**Reply by mario-website** [reply]
> last night the configurator worked..but today it doesn't work..it's very convenient if it works..thanks..16 Μαΐ 2023, 19:18, ο χρήστης «mario-website ***@***.***>» έγραψε:﻿ Very good idea the converter..but not working for me..😐 I learning Merlin... I believe this is related to [https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4] —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

Hi, thanks for trying using. Could you tell me more why is not working (what was working yesterday and what is not working today: show me a config file from yesterday and today) and any error messages, then I might be able to help.
**Please note, that this converter works for now ONLY IF some condition is meet.** An working example is to convert from Sonoff 4CH template form https://glumb.github.io/haa-configurator/.
Few minutes ago I just added a condition to include "l" value in array where is [[...allOtherValues],2]

**Reply by mario-website** [reply]
...there is so many condition to check in conversion but probably, one day all of them will be checked and converter will be working as expected.
Please ANYONE, share your code if this website helped you so I can use to check in testing area as examples of input and expected output as when this convert will pass more conditions, previous worked conditions might unintentionally failed.

**Reply by xrust83** [reply]
If the GPIO type for the LED is set to 2, then the LED can have two states, On and Off. If the GPIO type for the LED is set to 3, then it changes brightness but does not turn off completely.

https://www.youtube.com/watch?v=kAUqxAhyNVA

**Reply by mario-website** [reply]
> > прошлой ночью конфигуратор работал... но сегодня он не работает... это очень удобно, если он работает... спасибо..16 Μαΐ 2023, 19:18, ο χρήστης «mario-website _**@**_.***>» έγραψε: Очень хорошая идея конвертера... но не работает для меня... �Я изучаю Мерлина... Я считаю, что это связано с [https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4] —Ответьте на это электронное письмо напрямую, просмотрите его на GitHub или откажитесь от подписки. Вы получаете это, потому что прокомментировали. Идентификатор сообщения: _**@**_. ***>
> 
> If the GPIO type for the LED is set to 2, then the LED can have two states, On and Off. If the GPIO type for the LED is set to 3, then it changes brightness but does not turn on completely.
> 
> https://www.youtube.com/watch?v=kAUqxAhyNVA

THANKS!
I just fixed, please reload page and check if conversion is working correctly

**Reply by xrust83** [reply]
I think that both options are correct, and should be used as you like. There are problems that the program cannot solve. Today, such a task is with the artificial intelligence of the autopilot. If the car cannot stop and it needs to knock down one of the pedestrians, who should be knocked down by a cat or a person (the answer is a cat), if two people are young or old (the answer is old), and the dead-end question is you 50 years old and a child 10 should the car save you or the child? I think that with the LED the correct answer is both 2 and 3. Perhaps in the configurator you need to select the mode of the configurator for the LED, or explain to everyone why this is so. Or do it the way you like, and let the rest use your configurator, or all the same, they will force themselves to study and deal with MEPLHAA 😉

**Reply by mario-website** [reply]
I believe it should go to 3 because:
https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4

**Comment by pasparakis** [comment]
in my case your website doesn't open .. while yesterday it does ..https://converter-to-merlin.netlify.app/😐17 Μαΐ 2023, 13:30, ο χρήστης «mario-website ***@***.***>» έγραψε:﻿

last night the configurator worked..but today it doesn't work..it's very convenient if it works..thanks..16 Μαΐ 2023, 19:18, ο χρήστης «mario-website @.>» έγραψε:﻿ Very good idea the converter..but not working for me..😐 I learning Merlin... I believe this is related to [https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4] —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: @.>

Hi, thanks for trying using. Could you tell me more why is not working (what was working yesterday and what is not working today: show me a config file from yesterday and today) and any error messages, then I might be able to help.
Please note, that this converter works for now ONLY IF some condition is meet.

**Comment by pasparakis** [comment]
I can’t connect to your website from my local network.. I can connect only from mobile network..Any idea?Στάλθηκε από το iPhone μου17 Μαΐ 2023, 14:24, ο χρήστης «Γιάννης Πασπαράκης ***@***.***>» έγραψε:﻿in my case your website doesn't open .. while yesterday it does ..https://converter-to-merlin.netlify.app/😐17 Μαΐ 2023, 13:30, ο χρήστης «mario-website ***@***.***>» έγραψε:﻿

last night the configurator worked..but today it doesn't work..it's very convenient if it works..thanks..16 Μαΐ 2023, 19:18, ο χρήστης «mario-website @.>» έγραψε:﻿ Very good idea the converter..but not working for me..😐 I learning Merlin... I believe this is related to [https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-4] —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: @.>

Hi, thanks for trying using. Could you tell me more why is not working (what was working yesterday and what is not working today: show me a config file from yesterday and today) and any error messages, then I might be able to help.
Please note, that this converter works for now ONLY IF some condition is meet.

**Reply by mario-website** [reply]
Use correct url:
https://converter-to-merlin.netlify.app/
